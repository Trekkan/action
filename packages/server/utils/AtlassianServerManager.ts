import fetch from 'node-fetch'
import makeAppLink from './makeAppLink'
import AtlassianManager from 'parabol-client/utils/AtlassianManager'

interface AuthQueryParams {
  grant_type: 'authorization_code'
  code: string
  redirect_uri: string
}

interface RefreshQueryParams {
  grant_type: 'refresh_token'
  refresh_token: string
}

interface OAuth2Response {
  access_token: string
  refresh_token: string
  error: any
  scope: string
}

class AtlassianServerManager extends AtlassianManager {
  fetch = fetch
  static async init(code: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'authorization_code',
      code,
      redirect_uri: makeAppLink('auth/atlassian')
    })
  }

  static async refresh(refreshToken: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'refresh_token',
      refresh_token: refreshToken
    })
  }

  static async fetchToken(partialQueryParams: AuthQueryParams | RefreshQueryParams) {
    const queryParams = {
      ...partialQueryParams,
      client_id: process.env.ATLASSIAN_CLIENT_ID,
      client_secret: process.env.ATLASSIAN_CLIENT_SECRET
    }

    const uri = `https://auth.atlassian.com/oauth/token`

    const tokenRes = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(queryParams)
    })

    const tokenJson = (await tokenRes.json()) as OAuth2Response

    const {access_token: accessToken, refresh_token: refreshToken, error, scope} = tokenJson
    if (error) {
      throw new Error(`Atlassian: ${error}`)
    }
    const providedScope = scope.split(' ')
    const matchingScope =
      new Set([...AtlassianServerManager.SCOPE.split(' '), ...providedScope]).size ===
      providedScope.length
    if (!matchingScope) {
      throw new Error(`bad scope: ${scope}`)
    }
    return new AtlassianServerManager(accessToken, refreshToken)
  }

  refreshToken?: string

  constructor(accessToken: string, refreshToken?: string) {
    super(accessToken)
    this.refreshToken = refreshToken
  }
}

export default AtlassianServerManager
