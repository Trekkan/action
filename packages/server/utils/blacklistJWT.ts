import getBlacklistJWTKey from './getBlacklistJWTKey'
import {Threshold, SubscriptionChannel} from 'parabol-client/types/constEnums'
import {toEpochSeconds} from './epochTime'
import publish from './publish'
import getRedis from './getRedis'

const blacklistJWT = async (userId: string, iat: number, mutatorId?: string) => {
  const key = getBlacklistJWTKey(userId)
  const redis = getRedis()
  const expiresIn = iat - toEpochSeconds(new Date()) + Math.floor(Threshold.JWT_LIFESPAN / 1000)
  await redis.set(key, iat, 'EX', expiresIn)
  publish(SubscriptionChannel.NOTIFICATION, userId, 'InvalidateSessionsPayload', {}, {mutatorId})
}

export default blacklistJWT
