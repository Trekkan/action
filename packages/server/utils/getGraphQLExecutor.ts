import {ExecutionResult} from 'graphql'
import {ServerChannel} from 'parabol-client/types/constEnums'
import PubSubPromise from './PubSubPromise'

export interface PubSubPromiseResponse extends ExecutionResult {
  jobId: string
}

let pubsub: PubSubPromise<PubSubPromiseResponse>
const getGraphQLExecutor = () => {
  if (!pubsub) {
    pubsub = new PubSubPromise(
      ServerChannel.GQL_EXECUTOR_REQUEST,
      ServerChannel.GQL_EXECUTOR_RESPONSE
    )
  }
  return pubsub
}
export default getGraphQLExecutor
