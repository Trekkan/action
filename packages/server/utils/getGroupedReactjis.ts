import Reactji from '../database/types/Reactji'
import {IReactji} from 'parabol-client/types/graphql'

const getGroupedReactjis = (reactjis: Reactji[], viewerId: string, idPrefix: string) => {
  const agg = {}
  for (let i = 0; i < reactjis.length; i++) {
    const {id, userId} = reactjis[i]
    const guid = `${idPrefix}:${id}`
    const isViewerReactji = viewerId === userId
    const record = agg[guid]
    if (!record) {
      agg[guid] = {id: guid, count: 1, isViewerReactji}
    } else {
      record.count++
      record.isViewerReactji = record.isViewerReactji || isViewerReactji
    }
  }
  return Object.values(agg) as IReactji[]
}

export default getGroupedReactjis
