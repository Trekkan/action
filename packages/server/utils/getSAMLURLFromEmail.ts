import base64url from 'base64url'
import getSSODomainFromEmail from 'parabol-client/utils/getSSODomainFromEmail'
import getRethink from '../database/rethinkDriver'
import {URL} from 'url'

const urlWithRelayState = (url: string, isInvited?: boolean | null) => {
  if (!isInvited) return url
  const relayState = base64url.encode(JSON.stringify({isInvited: true}))
  const urlObj = new URL(url)
  urlObj.searchParams.append('RelayState', relayState)
  return urlObj.toString()
}

const getSAMLURLFromEmail = async (email: string, isInvited?: boolean | null) => {
  const domainName = getSSODomainFromEmail(email)
  if (!domainName) return null
  const r = await getRethink()
  const baseURL = (await r
    .table('SAML')
    .getAll(domainName, {index: 'domain'})
    .nth(0)('url')
    .default(null)
    .run()) as string | null
  if (!baseURL) return null
  return urlWithRelayState(baseURL, isInvited)
}

export default getSAMLURLFromEmail
