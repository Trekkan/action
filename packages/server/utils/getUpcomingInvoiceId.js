const getUpcomingInvoiceId = (orgId) => `upcoming_${orgId}`

export default getUpcomingInvoiceId
