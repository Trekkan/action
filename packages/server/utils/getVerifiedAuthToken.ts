import {verify} from 'jsonwebtoken'
import AuthToken from '../database/types/AuthToken'

const SERVER_SECRET_BUFFER = Buffer.from(process.env.SERVER_SECRET!, 'base64')

const getVerifiedAuthToken = (jwt: string | undefined | null, ignoreExp?: boolean) => {
  if (!jwt) return {} as AuthToken
  try {
    return verify(jwt, SERVER_SECRET_BUFFER, {ignoreExpiration: ignoreExp}) as AuthToken
  } catch (e) {
    return {} as AuthToken
  }
}

export default getVerifiedAuthToken
