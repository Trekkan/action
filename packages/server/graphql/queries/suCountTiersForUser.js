import {GraphQLID, GraphQLNonNull} from 'graphql'
import UserTiersCount from '../types/UserTiersCount'
import {requireSU} from '../../utils/authorization'
import countTiersForUserId from './helpers/countTiersForUserId'

export default {
  type: UserTiersCount,
  args: {
    userId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'the user for which you want the count of tier membership'
    }
  },
  async resolve (source, args, {authToken}) {
    const {userId} = args

    // AUTH
    requireSU(authToken)

    // RESOLUTION
    return {
      ...(await countTiersForUserId(userId)),
      userId
    }
  }
}
