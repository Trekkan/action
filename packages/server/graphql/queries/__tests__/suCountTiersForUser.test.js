/* eslint-env jest */
import mockAuthToken from '../../../__tests__/setup/mockAuthToken'
import MockDB from '../../../__tests__/setup/MockDB'
import expectAsyncToThrow from '../../../__tests__/utils/expectAsyncToThrow'
import suCountTiersForUser from '../suCountTiersForUser'
import { PRO } from 'parabol-client/utils/constants'

console.error = jest.fn()

test('return counts for a userId', async () => {
  // SETUP
  const mockDB = new MockDB()
  const { user } = await mockDB.init()
  const authToken = mockAuthToken(user[0], { rol: 'su' })
  // TEST
  const result = await suCountTiersForUser.resolve(undefined, { userId: user[0].id }, { authToken })

  // VERIFY
  expect(result.tierPersonalCount).toBeGreaterThanOrEqual(0)
  expect(result.tierProCount).toBeGreaterThanOrEqual(0)
  expect(result.tierProBillingLeaderCount).toBeGreaterThanOrEqual(0)
  expect(result.userId).toBe(user[0].id)
})

test('pro count increments on new Pro Org for userId', async () => {
  // SETUP
  const mockDB = new MockDB()
  await mockDB.init().newOrg({ name: 'Marvel', tier: PRO })
  const authToken = mockAuthToken(mockDB.context.user, { rol: 'su' })
  // TEST
  const result = await suCountTiersForUser.resolve(
    undefined,
    { userId: mockDB.context.user.id },
    { authToken }
  )

  // VERIFY
  expect(result.tierPersonalCount).toBeGreaterThanOrEqual(0)
  expect(result.tierProCount).toBeGreaterThanOrEqual(1)
  expect(result.tierProBillingLeaderCount).toBeGreaterThanOrEqual(1)
  expect(result.userId).toBe(mockDB.context.user.id)
})

test('user token requires su role', async () => {
  // SETUP
  const mockDB = new MockDB()
  const { user } = await mockDB.init()
  const authToken = mockAuthToken(user[0])

  // TEST & VERIFY
  await expectAsyncToThrow(
    suCountTiersForUser.resolve(undefined, { userId: user[0].id }, { authToken })
  )
})
