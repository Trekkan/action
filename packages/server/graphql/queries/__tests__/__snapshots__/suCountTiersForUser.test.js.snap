// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user token requires su role 1`] = `"Unauthorized. Must be a super user to run this query."`;
